openedx_tagging/core/tagging/system_defined_taxonomies/object_tags.py

```
@@ -0,0 +1,210 @@
+
+    Using this approach makes the connection between the ObjectTag
+    and system defined taxonomy as hardcoded and can't be changed.
+    """
+
+    system_defined_taxonomy_id = None
+
+    def _check_system_taxonomy(self, taxonomy: Taxonomy = None):
+        """
+        Validates if the taxonomy is system-defined and match
+        with the name stored in the object tag
+        """
+        return (
+            bool(taxonomy) and
+            taxonomy.system_defined and
+            taxonomy.id == self.system_defined_taxonomy_id
+        )
+
+
+class OpenSystemObjectTag(OpenObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Free-text object tag used on system-defined taxonomies
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+    
+class ClosedSystemObjectTag(ClosedObjectTag, SystemDefinedObjectTagMixin):
+    """
+    Object tags linked to a closed system-taxonomy
+    """
+
+    class Meta:
+        proxy = True
+
+    def _check_taxonomy(self):
+        return (
+            super()._check_taxonomy() and
+            self._check_system_taxonomy(self.taxonomy)
+        )
+
+
+class ModelObjectTag(OpenSystemObjectTag):
+    """
+    Object tags used with tags that relate to the id of a model
+    
+    This object tag class is not registered as it needs to have an associated model
+    """
+
+    class Meta:
+        proxy = True
+
+    tag_class_model = None
+
+
+    def _check_taxonomy(self):
+        """
+        Validates if has an associated Django model that has an Id
+        """
+        if not super()._check_taxonomy():
+            return False
+    
+        if not self.tag_class_model:
+            return False
+
+        # Verify that is a Django model
+        if not issubclass(self.tag_class_model, models.Model):
+            return False
```

This should raise an exception instead of just silently failing, to make it easier for developers to figure out what's wrong with their ModelObjectTag subclasses..

I'm not sure if this is the best place to raise this exception though, so am open to other ideas.

```suggestion
        # Verify that is a Django model
        assert issubclass(self.tag_class_model, models.Model)
```

---

[Updated here](https://github.com/open-craft/openedx-learning/pull/3/files#diff-204875f41064da3abfd4648611ad41c6f21abe09277263410e3daafbd343f2a3R687)